spec/reports/middleware/cache_spec.rb

```diff
@@ -0,0 +1,151 @@
+      stubs.get("http://example.test") { [404, {}, "not found"] }
+
+      response = conn.get "http://example.test"
+      expect(response.status).to eql(404)
+    end
+
+    it "does not cache when the response Cache-Controll header has no-store value" do
+      stubs.get("http://example.test") { [200, {'Cache-Control' => 'no-store'}, "hello"] }
+      conn.get("http://example.test")
+      stubs.get("http://example.test") { [404, {}, "not found"] }
+
+      response = conn.get "http://example.test"
+      expect(response.status).to eql(404)
+    end
+
+    it "does not use cached response when the response Cache-Controll header has no-cache value" do
+      stubs.get("http://example.test") { [200, {'Cache-Control' => 'no-store'}, "hello"] }
+      conn.get("http://example.test")
+      stubs.get("http://example.test") { [404, {}, "not found"] }
+
+      response = conn.get "http://example.test"
+      expect(response.status).to eql(404)
+    end
+
+    it "does not use cached response when the response Cache-Controll header has must-validate value" do
+      stubs.get("http://example.test") { [200, {'Cache-Control' => 'must-validate'}, "hello"] }
+      conn.get("http://example.test")
+      stubs.get("http://example.test") { [404, {}, "not found"] }
+
+      response = conn.get "http://example.test"
+      expect(response.status).to eql(404)
+    end
+
+    it "uses cached response when it doesn't exceeds max age" do
+      stubs.get("http://example.test") { [200, { 'Cache-Control' => 'max-age=60', 'Date' => Time.now.httpdate}, "hello"] }
+      conn.get("http://example.test")
+      stubs.get("http://example.test") { [404, {}, "not found"] }
+
+      response = conn.get "http://example.test"
+      expect(response.status).to eql(200)
+    end
+
+    it "does not use cached response when it does exceeds max age" do
+      stubs.get("http://example.test") { [200, { 'Cache-Control' => 'max-age=60', 'Date' => (Time.now - 2 * 60).httpdate }, "hello"] }
+
+      conn.get("http://example.test")
+      stubs.get("http://example.test") { [404, {}, "not found"] }
+
+      response = conn.get "http://example.test"
+      expect(response.status).to eql(404)
+    end
+
+    it "refetches a stale response and replaces it with a new one" do
+      # set up the cache for a response that's 2 minutes old
+      two_minutes_ago = Time.now - 2 * 60
+      cached_response_headers = { 'Cache-Control' => 'max-age=60', 'Date' => two_minutes_ago.httpdate, 'ETag' => 'oldETag' }
+      stubs.get("http://example.test") { [200, cached_response_headers, "hello"] }
+      conn.get("http://example.test")
+
+      # assert that a conditional request will be sent with `If-None-Match` header set, and the server responds with 200,
+      # with a new body since the state has changed on the server side
+
+      now_time = Time.now
+      response_headers = { 'Cache-Control' => 'max-age=60', 'Date' => now_time.httpdate, 'ETag' => 'newETag' }
```

Comment:
Line is too long. [110/80]<br>Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.